### PR TITLE
[TF] RNN

### PR DESCRIPTION
**Pull request by abhinavsp0730**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by abhinavsp0730** [comment]
please don't review now.

File: chapter_recurrent-neural-networks/rnn-scratch.md [hunk 1]
```diff
@@ -360,6 +489,34 @@ def train_epoch_ch8(model, train_iter, loss, updater, device,  #@save
...
+        with tf.GradientTape(persistent=True) as g:
+            g.watch(params)
+            py, state= model(X.numpy(), state, params)
+            y = d2l.reshape(Y.numpy().T, (-1))
+            l = tf.math.reduce_mean(loss(y, py)) 
+        grads = g.gradient(l, params)
+        grads = grad_clipping(grads, 1)
+        updater.apply_gradients(zip(grads, params))
+        
+        # Keras loss by default returns the average loss in a batch
```
Comment (abhinavsp0730) [inline chapter_recurrent-neural-networks/rnn-scratch.md:513]: @terrytangyuan  do I need l_sum or l here?
i.e metric.add(l* d2l.size(y), d2l.size(y)) or metric.add(l_sum* d2l.size(y), d2l.size(y))

**Comment by mli** [comment]
Job d2l-en/PR-1301/4 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1301/

**Comment by mli** [comment]
Job d2l-en/PR-1301/7 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1301/

**Comment by mli** [comment]
Job d2l-en/PR-1301/8 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1301/

**Comment by mli** [comment]
Job d2l-en/PR-1301/9 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1301/

**Comment by mli** [comment]
Job d2l-en/PR-1301/11 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1301/

**Comment by mli** [comment]
Job d2l-en/PR-1301/12 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1301/

**Comment by abhinavsp0730** [comment]
It's now ready to review.
Notes:
1. I've use one extra argument ```params``` in model().
I'm compelled to do so because in order to make params connected to 
to the loss within the tf.GradienTape() context.
If I don't do this, then the gradients will be none. 

2.  The perpexuility is too high, I've spend a lot of time debugging it. But not able to find what is causing this. What is happening under the hood is after some epochs the gradients are vanishing( the gradients are getting near to zero and then become zero).
Sometimes it may happens that the bug is not seen by the person who wrote the code. But it's easily catched by other person.

3. I've loaded two dataset, each is having different random_iter. Please have a look that I posted in #TensorFlow.

File: chapter_recurrent-neural-networks/rnn-scratch.md (hunk 1 above)
Comment (abhinavsp0730) [reply chapter_recurrent-neural-networks/rnn-scratch.md:528]: what about this?

**Comment by abhinavsp0730** [comment]
> Thanks. Please make sure to track all the remaining issues/TODOs somewhere, e.g. in the spreadsheet.

Thanks, added in the spreadsheet.

**Comment by AnirudhDagar** [comment]
@abhinavsp0730 please commit the whole `d2l/tensorflow` package by running `d2lbook build lib`. I think this commit is missing a few changes in d2l/tensorflow.
cc @astonzhang PTAL.

Please confirm this @abhinavsp0730.

This is the `diff` i got after running `d2lbook build lib` for `d2l/tensorflow.py` 
```
-    norm = tf.math.sqrt(tf.reduce_sum([tf.reduce_sum(grad ** 2)
-                                       for grad in grads]))
+    norm = tf.math.sqrt(sum((tf.reduce_sum(grad ** 2)).numpy()
+                        for grad in grads))
+    norm = tf.cast(norm, tf.float32)
```

**Comment by astonzhang** [comment]
> > Thanks. Please make sure to track all the remaining issues/TODOs somewhere, e.g. in the spreadsheet.
> 
> Thanks, added in the spreadsheet.

Can you also track them in the GitHub issues?

File: chapter_recurrent-neural-networks/rnn-scratch.md (hunk 1 above)
Comment (astonzhang) [reply chapter_recurrent-neural-networks/rnn-scratch.md:528]: @terrytangyuan can you take a look at this?

**Comment by astonzhang** [comment]
> @abhinavsp0730 please commit the whole `d2l/tensorflow` package by running `d2lbook build lib`. I think this commit is missing a few changes in d2l/tensorflow.
> cc @astonzhang PTAL.
> 
> Please confirm this @abhinavsp0730.
> 
> This is the `diff` i got after running `d2lbook build lib` for `d2l/tensorflow.py`
> 
> ```
> -    norm = tf.math.sqrt(tf.reduce_sum([tf.reduce_sum(grad ** 2)
> -                                       for grad in grads]))
> +    norm = tf.math.sqrt(sum((tf.reduce_sum(grad ** 2)).numpy()
> +                        for grad in grads))
> +    norm = tf.cast(norm, tf.float32)
> ```

fixed https://github.com/d2l-ai/d2l-en/commit/f75452da0c7b3d511599dcee8f6ee29df42e8888

**Comment by abhinavsp0730** [comment]
> > @abhinavsp0730 please commit the whole `d2l/tensorflow` package by running `d2lbook build lib`. I think this commit is missing a few changes in d2l/tensorflow.
> > cc @astonzhang PTAL.
> > Please confirm this @abhinavsp0730.
> > This is the `diff` i got after running `d2lbook build lib` for `d2l/tensorflow.py`
> > ```
> > -    norm = tf.math.sqrt(tf.reduce_sum([tf.reduce_sum(grad ** 2)
> > -                                       for grad in grads]))
> > +    norm = tf.math.sqrt(sum((tf.reduce_sum(grad ** 2)).numpy()
> > +                        for grad in grads))
> > +    norm = tf.cast(norm, tf.float32)
> > ```
> 
> fixed [f75452d](https://github.com/d2l-ai/d2l-en/commit/f75452da0c7b3d511599dcee8f6ee29df42e8888)

Sorry, it is happens by mistake. I thought it was just a text changes. And forget to run d2l build lib.

**Comment by abhinavsp0730** [comment]
Also please look at number 3 here: https://github.com/d2l-ai/d2l-en/pull/1301#issuecomment-671856068

What I did is added two version of dataset each having different  iter. Please take a look it's not done in this way in pt/mxnet.

**Comment by abhinavsp0730** [comment]
Also please look at number 3 here: https://github.com/d2l-ai/d2l-en/pull/1301#issuecomment-671856068

What I did is added two version of dataset each having different  iter. Please take a look it's not done in this way in pt/mxnet.
